Type hinting developer docs (#26051)
matplotlib/matplotlib: doc/devel/contribute.rst

   - Use ``_api.warn_deprecated()`` for general deprecation warnings
   - Use the decorator ``@_api.deprecated`` to deprecate classes, functions,
     methods, or properties

>    - Use the decorator ``@_api.deprecate_privatize_attribute`` to deprecate

Reviewer: This isn't a decorator; it's a fancy descriptor thing.

   You can use standard rst cross references in *alternative*.

3. Make appropriate changes to the type hints in the associated ``.pyi`` file.
   The general guideline is to match runtime reported behavior.

  - Items marked with ``@_api.deprecated`` or ``@_api.deprecate_privatize_attribute``
    are generally kept during the expiry period, and thus no changes are needed on
    introduction.
  - Items decorated with ``@_api.rename_parameter`` or ``@_api.make_keyword_only``
    report the *new* (post deprecation) signature at runtime, and thus *should* be
    updated on introduction.
  - Items decorated with ``@_api.delete_parameter`` should include a default value hint
    for the deleted parameter, even if it did not previously have one (e.g.
    ``param: <type> = ...``). Even so, the decorator changes the default value to a
    sentinel value which should not be included in the type stub. Thus, Mypy Stubtest
    needs to be informed of the inconsistency by placing the method into

>     ``ci/mypy-stubtest-allowlist.txt`` under a heading indicating the deprecation

Reviewer: ```suggestion
    :file:`ci/mypy-stubtest-allowlist.txt` under a heading indicating the deprecation
```

   information. For the content, you can usually copy the deprecation notice
   and adapt it slightly.
2. Change the code functionality and remove any related deprecation warnings.
3. Make appropriate changes to the type hints in the associated ``.pyi`` file.

  - Items marked with ``@_api.deprecated`` or ``@_api.deprecate_privatize_attribute``
    are to be removed on expiry.
  - Items decorated with ``@_api.rename_parameter`` or ``@_api.make_keyword_only``

>     will have been updated at introduction.

Reviewer: ```suggestion
    will have been updated at introduction, and require no change now.
```

  - Items decorated with ``@_api.delete_parameter`` will need to be updated to the
    final signature, in the same way as the ``.py`` file signature is updated.

>     The entry in ``ci/mypy-stubtest-allowlist.txt`` should be removed.
>   - Any other entries in ``ci/mypy-stubtest-allowlist.txt`` under a versions

Reviewer: ```suggestion
    The entry in :file:`ci/mypy-stubtest-allowlist.txt` should be removed.
  - Any other entries in :file:`ci/mypy-stubtest-allowlist.txt` under a version's
```
